### PR TITLE
Lazily download tools

### LINKED ISSUE #59: Unused tools are fetched
I'd expect that if I add an entry to multitool.lock.json which is not referenced anywhere, the corresponding toolchain will be defined but the tool itself is never fetched.

However https://github.com/theoremlp/rules_multitool/pull/58 shows a reproduction adding a new entry and it DOES get fetched. This means users with a big lockfile (or who accumulate third-party lockfiles into their hub, such as from rules_lint) will spend time fetching tools they don't need.

### PR DESCRIPTION
**Pull request by mark-thm**
Closes #59.

Refactor os/cpu repos into per-tool repos and split bzlmod/non-bzlmod handling in order to add a workspace-only shim. As a result of these changes, only tools needed by a build will be downloaded/all tool downloads become lazy.

**Breaking change** for workspace users.

### CODE REVIEW

File: examples/workspace/minimal_download_test.sh [hunk 1]
```diff
@@ -21,6 +21,7 @@ for file in target-determinator.darwin.arm64 target-determinator.linux.amd64 tar
   if [ "$file" != "$ALLOWED" ]; then
     cat >$DL_CONFIG <<EOF
 rewrite github.com/bazel-contrib/target-determinator/releases/download/([^/]+)/$file disallowed.build/\$1/$file
+rewrite github.com/cli/cli/releases/download/(.*) disallowed.build/\$1
```
Comment (mark-thm) [inline examples/workspace/minimal_download_test.sh:24]: this is cribbed from #58

File: multitool/private/multitool.bzl [hunk 2]
```diff
@@ -124,122 +126,128 @@ def _extension(os):
         return ".exe"
     return ""
 
-def _env_specific_tools_impl(rctx):
-    tools = _load_tools(rctx)
+def _download_extract_tool(rctx, tool_name, binary):
```
Comment (mark-thm) [inline multitool/private/multitool.bzl:129]: this is the body of the double loop from old `env_specific_tools_impl`

File: multitool/private/templates.bzl [hunk 3]
```diff
@@ -38,9 +40,40 @@ def _render_hub_tool(rctx, tool_name, filename, substitutions = None):
...
+        os = binary["os"],
+        cpu = binary["cpu"],
+    )
+    return "\n".join([
+        "    tool_repo(",
+        "        name = \"{name}\",".format(name = name),
+        "        tool_name = \"{tool_name}\",".format(tool_name = tool_name),
+        "        binary = '{binary}',".format(binary = json.encode(binary)),
+        "    )",
+        "",
```
Comment (mark-thm) [inline multitool/private/templates.bzl:57]: there's probably a nicer way to do some of this templating but this will do for now

**Review by kevans-theorem — APPROVED** [review]

**Comment by alexeagle** [comment]
Awesome thanks!!

**Pull request merged**